Make hits on overflow controls target RenderTextControlMultiline itself

fantasai opened the pull request:
#### 00e7f006e4094af8428002a407198255336fcceb
<pre>
Make hits on overflow controls target RenderTextControlMultiline itself
<a href="https://bugs.webkit.org/show_bug.cgi?id=266873">https://bugs.webkit.org/show_bug.cgi?id=266873</a>
<a href="https://rdar.apple.com/problem/120108102">rdar://problem/120108102</a>

Reviewed by Simon Fraser.

When align-content moves the textarea contents, the coordinates of the
inner text block no longer match those of the RenderTextControlMultiline,
so hit tests using the inner block&apos;s coordinates don&apos;t land on the resizer
properly. This patch captures these cases and associates them with the
textarea, rather than assuming they belong to the inner text block.

Side-effects of this patch include:
- Capturing hits correctly in the presence of borders/padding.
  <a href="https://bugs.webkit.org/show_bug.cgi?id=238193">https://bugs.webkit.org/show_bug.cgi?id=238193</a> &amp; <a href="https://rdar.apple.com/90639221">rdar://90639221</a>
- Using the correct cursor when hovering over the resize controls.
- Not jumping by the border-width when triggering a resize operation.
Test expectations on Linux are thus updated accordingly.

* Source/WebCore/rendering/RenderTextControlMultiLine.cpp:
(WebCore::RenderTextControlMultiLine::nodeAtPoint): Capture hits in overflow controls.
* LayoutTests/fast/css/resize-textarea-align-content-expected.txt: Added.
* LayoutTests/fast/css/resize-textarea-align-content.html: Added.
* LayoutTests/platform/gtk/fast/css/resize-corner-tracking-expected.txt: Use corrected expectations.
* LayoutTests/platform/ios/TestExpectations: Skip new test due to no resize support.
* LayoutTests/platform/wpe/fast/css/resize-corner-tracking-expected.txt: Use corrected expectations.

Canonical link: <a href="https://commits.webkit.org/272964@main">https://commits.webkit.org/272964@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/12342611fc0b6dee85c6b5299565280d21f185f3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/33718 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/12491 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/35652 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/36333 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/30571 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/34766 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/14863 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/9644 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/29648 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/34195 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/10532 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/30060 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/9189 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/9284 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/30074 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/37656 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/30600 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/30394 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/35416 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/9425 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/7370 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/33304 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/11205 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/29735 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7791 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/10008 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/10211 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/rendering/RenderTextControlMultiLine.cpp:
```diff
@@ -58,6 +59,11 @@ bool RenderTextControlMultiLine::nodeAtPoint(const HitTestRequest& request, HitT
     if (!RenderTextControl::nodeAtPoint(request, result, locationInContainer, accumulatedOffset, hitTestAction))
         return false;
 
+    const LayoutPoint localPoint(locationInContainer.point() - (accumulatedOffset + location()));
```
I think it would be better to implement `isPointInOverflowControl()` like RenderBlock and RenderListBox do and have it call `hitTestOverflowControls()`, just for consistency with those other classes.

fantasai replied:
OK, I moved `isPointInOverflowControl()` from `private` to `protected` and called it.

fantasai committed:
Make hits on overflow controls target RenderTextControlMultiline itself
https://bugs.webkit.org/show_bug.cgi?id=266873
rdar://problem/120108102

Reviewed by Simon Fraser.

When align-content moves the textarea contents, the coordinates of the
inner text block no longer match those of the RenderTextControlMultiline,
so hit tests using the inner block's coordinates don't land on the resizer
properly. This patch captures these cases and associates them with the
textarea, rather than assuming they belong to the inner text block.

Side-effects of this patch include:
- Capturing hits correctly in the presence of borders/padding.
  https://bugs.webkit.org/show_bug.cgi?id=238193 & rdar://90639221
- Using the correct cursor when hovering over the resize controls.
- Not jumping by the border-width when triggering a resize operation.
Test expectations on Linux are thus updated accordingly.

* Source/WebCore/rendering/RenderTextControlMultiLine.cpp:
(WebCore::RenderTextControlMultiLine::nodeAtPoint): Capture hits in overflow controls.
* LayoutTests/fast/css/resize-textarea-align-content-expected.txt: Added.
* LayoutTests/fast/css/resize-textarea-align-content.html: Added.
* LayoutTests/platform/gtk/fast/css/resize-corner-tracking-expected.txt: Use corrected expectations.
* LayoutTests/platform/ios/TestExpectations: Skip new test due to no resize support.
* LayoutTests/platform/wpe/fast/css/resize-corner-tracking-expected.txt: Use corrected expectations.

Canonical link: https://commits.webkit.org/272964@main
